### PR TITLE
feat: support compressed image

### PR DESCRIPTION
**Pull request by wep21**
support conversion of compressed image

### CODE REVIEW

**Comment by wep21** [comment]
@ivanpauno Could you review this PR?

File: rosbag2_to_video/api.py [hunk 1]
```diff
@@ -58,17 +78,20 @@ def convert_bag_to_video(args):
...
+    if type_map[args.topic] == 'sensor_msgs/msg/Image':
+        cv_image = cvbridge.imgmsg_to_cv2(image_msg, 'bgr8')
+    elif type_map[args.topic] == 'sensor_msgs/msg/CompressedImage':
+        cv_image = cvbridge.compressed_imgmsg_to_cv2(image_msg, 'bgr8')
     height, width, _ = cv_image.shape
     success = video_writer.open(
         args.output,
         cv2.CAP_FFMPEG,
-        cv2.VideoWriter.fourcc('a', 'v', 'c', '1'),
+        cv2.VideoWriter.fourcc('m', 'p', '4', 'v'),
```
Comment (ivanpauno) [inline rosbag2_to_video/api.py]: please don't change the default encoding, H264 is a good default.
If you want to add an option to select another encoding, that's fine

File: rosbag2_to_video/api.py [hunk 2]
```diff
@@ -30,25 +30,45 @@ def get_stamp_from_image_msg(image_msg):
 
 def add_arguments_to_parser(argparser):
     argparser.add_argument(
-        'bagfile', help='Path to the bag (currently only supports sensor_msgs/msg/Image types)')
-    argparser.add_argument('-t','--topic', required=True, help='Name of the image topic')
-    argparser.add_argument('-o','--output', required=True, help='Output filename (currently only supports MP4)')
+        'bagfile',
+        help='Path to the bag (currently only supports sensor_msgs/msg/Image or sensor_msgs/msg/CompressedImage types)' # noqa E501
```
Comment (ivanpauno) [inline rosbag2_to_video/api.py]: could you move the comment about the supported types to `-t/--topic`?
Thanks!

File: rosbag2_to_video/api.py [hunk 3]
```diff
@@ -30,25 +30,45 @@ def get_stamp_from_image_msg(image_msg):
...
+        help='Name of the image topic'
+    )
+    argparser.add_argument(
+        '-o',
+        '--output',
+        required=True,
+        help='Output filename (currently only supports MP4)'
+    )
     argparser.add_argument('--fps', type=int, required=True, help='Output frames per second')
+    argparser.add_argument('--storage-id', type=str, default="sqlite3", help='Rosbag2 storage id')
```
Comment (ivanpauno) [inline rosbag2_to_video/api.py:52]: why is this needed?
can this be read from the rosbag2 metadata?

File: rosbag2_to_video/api.py [hunk 4]
```diff
@@ -30,25 +30,45 @@ def get_stamp_from_image_msg(image_msg):
...
 
     bag_reader = rosbag2_py.SequentialReader()
     bag_reader.open(storage_options, converter_options)
 
+    topic_types = bag_reader.get_all_topics_and_types()
+    type_map = {topic_types[i].name: topic_types[i].type for i in range(len(topic_types))}
+    print(type_map[args.topic])
+    if (type_map[args.topic] != 'sensor_msgs/msg/Image' and
+            type_map[args.topic] != 'sensor_msgs/msg/CompressedImage'):
+        raise ValueError('topic type should be sensor_msgs/msg/Image or sensor_msgs/msg/CompressedImage') # noqa E501
```
Comment (ivanpauno) [inline rosbag2_to_video/api.py]: You need the topic type for `args.topic`, not for anything else.
So constructing a map seems unnecessary.

```suggestion
    try:
        topic_type = next(x for x in topic_types if x.name == args.topic).type
    except StopIteration:
        print(f'Topic {args.topic} was not recorded in the bagfile')
        return
    if topic_type not in ('sensor_msgs/msg/Image', 'sensor_msgs/msg/CompressedImage'):
          print(
              'topic type should be sensor_msgs/msg/Image or '
              f'sensor_msgs/msg/CompressedImage, got {topic_type}')
```

File: rosbag2_to_video/api.py [hunk 5]
```diff
@@ -58,17 +78,20 @@ def convert_bag_to_video(args):
     if not bag_reader.has_next():
         print('empty bag file')
         return
-    msg_type = get_message('sensor_msgs/msg/Image')
+    msg_type = get_message(type_map[args.topic])
```
Comment (ivanpauno) [inline rosbag2_to_video/api.py]:
```suggestion
    msg_type = get_message(topic_type)
```

Comment (ivanpauno) [reply rosbag2_to_video/api.py]: same elsewhere

File: rosbag2_to_video/api.py [hunk 6]
```diff
@@ -89,8 +112,11 @@ def convert_bag_to_video(args):
         msg_count += 1
         _, data, _ = bag_reader.read_next()
         image_msg = deserialize_message(data, msg_type)
-        # cv2.VideoWriter expects BGR encoding 
-        cv_image = cvbridge.imgmsg_to_cv2(image_msg, 'bgr8')
+        # cv2.VideoWriter expects BGR encoding
+        if type_map[args.topic] == 'sensor_msgs/msg/Image':
+            cv_image = cvbridge.imgmsg_to_cv2(image_msg, 'bgr8')
+        elif type_map[args.topic] == 'sensor_msgs/msg/CompressedImage':
+            cv_image = cvbridge.compressed_imgmsg_to_cv2(image_msg, 'bgr8')
```
Comment (ivanpauno) [inline rosbag2_to_video/api.py]: the repetition is fine for the moment, but I will refactor the code later to avoid repetition

File: rosbag2_to_video/api.py (hunk 3 above)
Comment (wep21) [reply rosbag2_to_video/api.py:52]: It is necessary if a user specify not a bag directory but a single db file as uri.

Comment (ivanpauno) [reply rosbag2_to_video/api.py:52]: I think requesting a directory instead of a single db is fine.
What does that happen if you specify a directory and also provide a storage id to `StorageOptions`?

If it figures it out correctly based on the metadata, then I think this change is fine.
But it would be nice to clarify that you don't need to pass the argument if you provide a directory, even if the storage id is different from `sqlite3`

File: rosbag2_to_video/api.py (hunk 2 above)
Comment (wep21) [reply rosbag2_to_video/api.py]: @ivanpauno I moved the comment in https://github.com/ivanpauno/rosbag2_to_video/pull/1/commits/741b1e325878d7a65f413fba7193859bf5646778.

File: rosbag2_to_video/api.py [hunk 7]
```diff
@@ -30,25 +30,49 @@ def get_stamp_from_image_msg(image_msg):
...
-    argparser.add_argument('-t','--topic', required=True, help='Name of the image topic')
-    argparser.add_argument('-o','--output', required=True, help='Output filename (currently only supports MP4)')
+        'bagfile',
+        help='Path to the bag'
+    )
+    argparser.add_argument(
+        '-t',
+        '--topic',
+        required=True,
+        help='Name of the image topic (currently only supports sensor_msgs/msg/Image or sensor_msgs/msg/CompressedImage types)' # noqa E501
```
Comment (ivanpauno) [inline rosbag2_to_video/api.py]: Avoid using `noqa`, except when strictly necessary

```suggestion
        help=(
            'Name of the image topic (currently only supports sensor_msgs/msg/Image or '
            'sensor_msgs/msg/CompressedImage types)')
```

File: rosbag2_to_video/api.py (hunk 1 above)
Comment (wep21) [reply rosbag2_to_video/api.py]: I added an argument for video codec. The default value is avc1. https://github.com/ivanpauno/rosbag2_to_video/pull/1/commits/595932065e54c9ac92b164b08a54ab8f4b51b195

File: rosbag2_to_video/api.py [hunk 8]
```diff
@@ -30,25 +30,50 @@ def get_stamp_from_image_msg(image_msg):
...
+    topic_types = bag_reader.get_all_topics_and_types()
+    try:
+        topic_type = next(x for x in topic_types if x.name == args.topic).type
+    except StopIteration:
+        print(f'Topic {args.topic} was not recorded in the bagfile')
+        return
+    if topic_type not in ('sensor_msgs/msg/Image', 'sensor_msgs/msg/CompressedImage'):
+          print(
+              'topic type should be sensor_msgs/msg/Image or '
+              f'sensor_msgs/msg/CompressedImage, got {topic_type}')
```
Comment (ivanpauno) [inline rosbag2_to_video/api.py:81]:
```suggestion
    if topic_type not in ('sensor_msgs/msg/Image', 'sensor_msgs/msg/CompressedImage'):
          print(
              'topic type should be sensor_msgs/msg/Image or '
              f'sensor_msgs/msg/CompressedImage, got {topic_type}')
          return
```

File: rosbag2_to_video/api.py [hunk 9]
```diff
@@ -30,25 +30,50 @@ def get_stamp_from_image_msg(image_msg):
...
+    )
+    argparser.add_argument(
+        '-o',
+        '--output',
+        required=True,
+        help='Output filename (currently only supports MP4)'
+    )
     argparser.add_argument('--fps', type=int, required=True, help='Output frames per second')
+    argparser.add_argument('--storage-id', type=str, default="sqlite3", help='Rosbag2 storage id')
+    argparser.add_argument('--codec', type=str, default="avc1", help='Video codec')
```
Comment (ivanpauno) [inline rosbag2_to_video/api.py:53]: could you check that the provided argument has 4 characters?
it would also be nice to have an option to list supported codecs, but that's out of the scope of this PR

**Review by ivanpauno — COMMENTED** [review]
Looks pretty good!

I have some minor comments, and I have to test it before merging.

File: rosbag2_to_video/api.py (hunk 3 above)
Comment (wep21) [reply rosbag2_to_video/api.py:52]: I added check if rosbag uri is a directory. https://github.com/ivanpauno/rosbag2_to_video/pull/1/commits/84034c699a47813b43db0fc1880704ef1d668de8

File: rosbag2_to_video/api.py (hunk 1 above)
Comment (wep21) [reply rosbag2_to_video/api.py]: I found out that avc1 and h264 is not supported in opencv python shipped by pip or apt due to the license issue.
https://stackoverflow.com/questions/59290569/opencv-video-writer-unable-to-find-codec-or-avc1
https://github.com/opencv/opencv-python/issues/207

**Review by ivanpauno — APPROVED** [review]
Thanks for the contribution
